test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NestedJSClassTest.scala

```diff
@@ -599,3 +620,20 @@ object NestedJSClassTest {
   }
 
 }
+
+object NestedJSClassTest_TopLevelJSObject_Issue4086 extends js.Object {
+  object InnerScalaObject {
+    override def toString(): String = "the InnerScalaObject of issue 4086"
+  }
+
+  object InnerJSObject extends js.Object {
+    override def toString(): String = "the InnerJSObject of issue 4086"
+  }
+
+  // Not visible
+  class InnerScalaClass
```

Comment:
I've added a test for the Scala-class-in-JS-class case. It behaves the same.

It is a bit odd, but I don't think we can do anything about it at this point. We can't make it *work*, because that would mean automatically *exporting* all non-private constructors of said class, which is not something we do anywhere else, and it could cause compilation failures for existing working code.

We could potentially warn for non-private Scala classes in JS objects and classes, although that should be discussed separately, I believe.

Reply:
> We could potentially warn for non-private Scala classes in JS objects and classes, although that should be discussed separately, I believe.

Agreed. Personally I think we should warn, but that's indeed a separate discussion.